data/transactions/logic/eval.go

```diff
@@ -2549,15 +2699,29 @@ func (cx *EvalContext) globalFieldToValue(fs globalFieldSpec) (sv stackValue, er
 	case LatestTimestamp:
 		sv.Uint, err = cx.getLatestTimestamp()
 	case CurrentApplicationID:
-		sv.Uint, err = cx.getApplicationID()
+		sv.Uint = uint64(cx.appID)
 	case CurrentApplicationAddress:
-		var addr basics.Address
-		addr, err = cx.getApplicationAddress()
+		addr := cx.getApplicationAddress(cx.appID)
 		sv.Bytes = addr[:]
 	case CreatorAddress:
 		sv.Bytes, err = cx.getCreatorAddress()
 	case GroupID:
-		sv.Bytes = cx.getGroupID()
+		sv.Bytes = cx.Txn.Txn.Group[:]
+	case OpcodeBudget:
+		sv.Uint = uint64(cx.remainingBudget())
+	case CallerApplicationID:
+		if cx.caller != nil {
+			sv.Uint = uint64(cx.caller.appID)
```

Comment:
interesting.. allowing to create private functions ;-)

Reply:
I was mostly worried about applications that do not wish to be invoked as inner transactions (like the bad random number generator app).  If such an application wanted to move to v6, but still not be invokable as an inner, they could check this and fail.

But I suppose an app could also refuse to run unless called by specific apps.